src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java

```diff
@@ -714,11 +714,13 @@ private void initalizeAlleleAnyploidIndicesCache(final VariantContext vc) {
                 // Make a new entry if the we have not yet cached a PL to allele indices map for this ploidy and allele count
                 // skip if there are no PLs -- this avoids hanging on high-allelic somatic samples, for example, where
                 // there's no need for the PL indices since they don't exist
-                if (g.getPloidy() != 0 && (!ploidyToNumberOfAlleles.containsKey(g.getPloidy()) || ploidyToNumberOfAlleles.get(g.getPloidy()) < vc.getNAlleles())) {
-                    if (vc.getGenotypes().stream().anyMatch(Genotype::hasLikelihoods)) {
-                        GenotypeLikelihoods.initializeAnyploidPLIndexToAlleleIndices(vc.getNAlleles() - 1, g.getPloidy());
-                        ploidyToNumberOfAlleles.put(g.getPloidy(), vc.getNAlleles());
-                    }
+                final int genotypePloidy = g.getPloidy();
+                final int nAlleles = vc.getNAlleles();
+                if (genotypePloidy != 0
+                        && (!ploidyToNumberOfAlleles.containsKey(genotypePloidy) || ploidyToNumberOfAlleles.get(genotypePloidy) < nAlleles)
+                        && g.hasLikelihoods()) {
```

Comment:
If you want to exactly match the previous behavior while still getting rid of the O(n^2) complexity, you could cache the result of `vc.getGenotypes().stream().anyMatch(Genotype::hasLikelihoods)` before entering the enclosing for loop, and check it here.

Reply:
(Otherwise the behavior will differ from the previous behavior if some genotypes have PLs and others don't)

Reply:
Oh, you're right.  I was thinking that since its just building up a cache it would hit it anyway when it got to one with PLs, but if the ploidy is different it will end up missing some.  Good call.

Reply:
Thanks for catching that @droazen. Probably best to preserve behavior. LGTM otherwise, thanks @lbergelson!

Reply:
@droazen  Instead of caching I moved it up to the outer guard.  Why loop at all if we're always going to skip.  Could you take a peak just to verify that that makes sense to you.